### PR TITLE
Replace the statefulset if it cannot be updated.

### PR DESCRIPTION
**Pull request by alexeyklyukin**
Updates to statefulset spec for fields other than 'replicas' and
containers' are forbidden. However, it is possible to delete the old
statefulset without deleting its pods and create the new one, using the
changed specs. The new statefulset shall pick up the orphaned pods.

Change the statefulset's comparison to return the combined effect of
all checks, not just the first non-matching field.

The environment variables in the statefulset are now sorted based on the name.
As a result, a one-time upgrade of all clusters might be necessary.

### CODE REVIEW

File: pkg/util/constants/constants.go [hunk 1]
```diff
@@ -1,31 +1,34 @@
...
+	ZalandoDnsNameAnnotation   = "external-dns.alpha.kubernetes.io/hostname"
+	ElbTimeoutAnnotationName   = "service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout"
+	ElbTimeoutAnnotationValue  = "3600"
+	KubeIAmAnnotation          = "iam.amazonaws.com/role"
+	ResourceName               = TPRName + "s"
+	PodRoleMaster              = "master"
+	PodRoleReplica             = "replica"
+	SuperuserKeyName           = "superuser"
+	ReplicationUserKeyName     = "replication"
+	RoleFlagSuperuser          = "SUPERUSER"
```
Comment (ikitiki) [inline pkg/util/constants/constants.go]: I'd add empty line before ``RoleFlagSuperuser`` and move ``StatefulsetDeletionTimeout`` just after ``ReplicationUserKeyName``

File: pkg/cluster/resources.go [hunk 2]
```diff
@@ -143,6 +146,56 @@ func (c *Cluster) updateStatefulSet(newStatefulSet *v1beta1.StatefulSet) error {
...
+	options := v1.DeleteOptions{OrphanDependents: &OrphanDepencies}
+	if err := c.KubeClient.StatefulSets(oldStatefulset.Namespace).Delete(oldStatefulset.Name, &options); err != nil {
+		return fmt.Errorf("Can't delete statefulset '%s': %s", statefulSetName, err)
+	}
+	// make sure we clear the stored statefulset status if the subsequent create fails.
+	c.Statefulset = nil
+	// wait until the statefulset is truly deleted
+	c.logger.Debugf("Waiting for the statefulset to be deleted")
+
+	before := time.Now()
```
Comment (ikitiki) [inline pkg/cluster/resources.go]: can we use ``retryutil.Retry()`` here?

Comment (alexeyklyukin) [inline pkg/cluster/resources.go]: Good idea, thanks.

File: pkg/cluster/resources.go [hunk 3]
```diff
@@ -2,12 +2,12 @@ package cluster
 
 import (
 	"fmt"
-	"time"
 
 	"k8s.io/client-go/pkg/api"
 	"k8s.io/client-go/pkg/api/v1"
 	"k8s.io/client-go/pkg/apis/apps/v1beta1"
 
+	"github.bus.zalan.do/acid/postgres-operator/pkg/util/retryutil"
```
Comment (ikitiki) [inline pkg/cluster/resources.go]: 😱
``github.com/zalando-incubator/postgres-operator/pkg/util/retryutil``

**Comment by Jan-M** [comment]
Can we double check and document our EBS / Persistent Volume reclaim settings? I would like to make sure that we really are doing this right when doing deletes outside of "deletes" triggered by deletes on the PostgreSQL object.

**Comment by alexeyklyukin** [comment]
@Jan-M I can't parse your comment. What does PV reclaim have to do with this pull request?

File: pkg/cluster/resources.go (hunk 3 above)
Comment (alexeyklyukin) [inline pkg/cluster/resources.go]: Ups, or IDE came out to be too smart. Fixed.

File: pkg/cluster/resources.go [hunk 4]
```diff
@@ -146,6 +149,53 @@ func (c *Cluster) updateStatefulSet(newStatefulSet *v1beta1.StatefulSet) error {
...
+func (c *Cluster) replaceStatefulSet(newStatefulSet *v1beta1.StatefulSet) error {
+	if c.Statefulset == nil {
+		return fmt.Errorf("There is no StatefulSet in the cluster")
+	}
+
+	statefulSetName := util.NameFromMeta(c.Statefulset.ObjectMeta)
+	c.logger.Debugf("Replacing StatefulSet")
+
+	// Delete the current statefulset without deleting the pods
+	OrphanDepencies := true
```
Comment (ikitiki) [inline pkg/cluster/resources.go]: i'd make it start with a lowercase letter

File: pkg/cluster/resources.go [hunk 5]
```diff
@@ -146,6 +149,53 @@ func (c *Cluster) updateStatefulSet(newStatefulSet *v1beta1.StatefulSet) error {
...
+	if err != nil {
+		fmt.Errorf("could not delete statefulset: %s", err)
+	}
+
+	// create the new statefulset with the desired spec. It would take over the remaining pods.
+	createdStatefulset, err := c.KubeClient.StatefulSets(newStatefulSet.Namespace).Create(newStatefulSet)
+	if err != nil {
+		return fmt.Errorf("Can't create statefulset '%s': %s", statefulSetName, err)
+	} else {
+		// check that all the previous replicas were picked up.
```
Comment (ikitiki) [inline pkg/cluster/resources.go]: the idea was to get the StatefulSet object from the Kubernetes and check if number of desired replicas match current number

**Comment by ikitiki** [comment]
👍

**Comment by alexeyklyukin** [comment]
👍

**Pull request merged**